Ball now moves and bounces off edges.

IvarForkbeard opened the pull request:
As requested, the ball now has a variable velocity and bounces off the edges of the canvas.

IvarForkbeard committed:
Ball moves and bounces off edges.

eric-decompiled commented on main.js:
```diff
@@ -46,6 +48,14 @@ const animate = () => {
   drawPaddles();
   drawBall();
   updatePaddles();
+  ball.x += ballSpeedHorizontal;
```
This should be in the ball class `update` method.

This animation function is for high level coordination processing a frame. 

Including low lever details like increment a number is a different level of abstraction then `drawBall`. This animate function should only tell the ball to update.

eric-decompiled commented on main.js:
```diff
@@ -46,6 +48,14 @@ const animate = () => {
   updatePaddles();
+  ball.x += ballSpeedHorizontal;
+  ball.y += ballSpeedVertical;
+  if (ball.x < 10 || ball.x > 949) {
```
`949` is acting as a "magic number" here. Its related to the other magic number we set the canvas width too.

This value should get a `const canvasWidth = 960` value near the top. Then the canvas width should be set to that value.

This makes it easier to update the size without having to search through the code.

eric-decompiled replied:
This should be `ball.width/2` to minimize "change amplification" of updating the ball size

eric-decompiled commented on main.js:
```diff
@@ -46,6 +48,14 @@ const animate = () => {
+  if (ball.x < 10 || ball.x > 949) {
+    ballSpeedHorizontal *= -1;
+  }
+  if (ball.y < 10 || ball.y > 539) {
```
Same comment on the width apply here. Use const for canvas size and calculate offset using the dimensions of the ball.

eric-decompiled commented on main.js:
```diff
@@ -46,6 +48,14 @@ const animate = () => {
   updatePaddles();
+  ball.x += ballSpeedHorizontal;
+  ball.y += ballSpeedVertical;
+  if (ball.x < 10 || ball.x > 949) {
+    ballSpeedHorizontal *= -1;
+  }
+  if (ball.y < 10 || ball.y > 539) {
+    ballSpeedVertical *= -1;
+  }
```
Unlike updating the position of the ball, which can be cleanly contained in the ball class, this is an interaction between two object, so has to be brought up a level in the abstraction tree.

Since this relates to the ball position it should go in the `updateBall` function already defined and called

IvarForkbeard committed:
Moved ball animations to different abstraction level.

Moved the ball animations from the animate level to the update ball level of abstraction.  Replaced hard coded canvas sizes in the ball updates with variables.

eric-decompiled commented on main.js:
```diff
@@ -38,6 +40,14 @@ const updatePaddles = () => {
 const updateBall = () => {
+  ball.x += ballSpeedHorizontal;
+  ball.y += ballSpeedVertical;
+  if (ball.x < ball.radius || ball.x > canvas.width / devicePixelRatio - ball.radius) {
```
```suggestion
  if (ball.x < ball.radius || ball.x > canvas.width - ball.radius) {
```

eric-decompiled commented on main.js:
```diff
@@ -38,6 +40,14 @@ const updatePaddles = () => {
+  if (ball.x < ball.radius || ball.x > canvas.width / devicePixelRatio - ball.radius) {
+    ballSpeedHorizontal *= -1;
+  }
+  if (ball.y < ball.radius || ball.y > canvas.height / devicePixelRatio - ball.radius) {
```
```suggestion
  if (ball.y < ball.radius || ball.y > canvas.height - ball.radius) {
```

eric-decompiled reviewed (commented):
the pixel ratio is already calculated in, doesn't look right when its being divided on my machine

eric-decompiled commented on main.js:
```diff
@@ -38,6 +40,14 @@ const updatePaddles = () => {
+  if (ball.y < ball.radius || ball.y > canvas.height / devicePixelRatio - ball.radius) {
+    ballSpeedVertical *= -1;
+  }
   ball.update();
```
I like how you have all the update code in this function, although its different from how I usually do it.

I put in update into the classes assuming they would contain their own velocity.

If you keep the velocity here, then there is no need to have an update and this can get cleaned up. (along with the class methods).

```suggestion
```

eric-decompiled replied:
don't worry about this if it doesn't make sense to you. its OK to leave the line as is for now

eric-decompiled reviewed (commented):
Short article on what the ratio is: https://cloudinary.com/glossary/device-pixel-ratio

If anything you will want to multiple your speed by the ratio if you want it consistent across devices. I suspect the ball appears slower on my screen compared to yours.

IvarForkbeard committed:
Removed pixel references to fix scaling issues.
